packages/fork-choice/src/protoArray/protoArray.ts

```diff
@@ -214,6 +232,135 @@ export class ProtoArray {
+
+    // Look reverse because its highly likely node with latestValidExecHash is towards the
+    // the leaves of the forkchoice
+    const latestValidHashIndexReverse = this.nodes
+      .slice()
+      .reverse()
+      .findIndex((node) => node.executionPayloadBlockHash === latestValidExecHash);
+    const latestValidHashIndex =
+      latestValidHashIndexReverse >= 0
+        ? this.nodes.length - 1 - latestValidHashIndexReverse
+        : latestValidHashIndexReverse;
+    if (latestValidHashIndex >= 0) {
+      this.propagateValidExecutionStatusByIndex(latestValidHashIndex);
+    }
+
+    if (invalidateTillBlockHash !== null) {
+      const invalidateTillBlockIndex = this.indices.get(invalidateTillBlockHash);
+      if (invalidateTillBlockIndex === undefined) {
+        throw Error(`Unable to find invalidateTillBlockHash=${invalidateTillBlockHash} in forkChoice`);
+      }
+      this.propagateInValidExecutionStatusByIndex(invalidateTillBlockIndex, latestValidHashIndex);
+      return;
+    }
+  }
+
+  propagateValidExecutionStatusByIndex(validNodeIndex: number): void {
+    let node: ProtoNode | undefined = this.getNodeFromIndex(validNodeIndex);
+    // propagate till we keep encountering syncing status
+    while (
+      node !== undefined &&
+      node.executionStatus !== ExecutionStatus.Valid &&
+      node.executionStatus !== ExecutionStatus.PreMerge
+    ) {
+      switch (node?.executionStatus) {
+        // If parent's execution is Invalid, we can't accept this block
+        // It should only happen on the leaf's parent, because we can't
+        // have a non invalid's parent as valid, if this happens it
+        // is a critical failure. So a chec
+        case ExecutionStatus.Invalid:
+          throw new ProtoArrayError({
+            code: ProtoArrayErrorCode.INVALID_PARENT_EXECUTION_STATUS,
+            root: node.blockRoot,
+            parent: node.parentRoot,
+          });
+
+        case ExecutionStatus.Syncing:
+          // If the node is valid and its parent is marked syncing, we need to
+          // propagate the execution status up
+          node.executionStatus = ExecutionStatus.Valid;
+          break;
+
+        default:
+      }
+
+      node = node.parent !== undefined ? this.getNodeByIndex(node.parent) : undefined;
+    }
+  }
+
+  propagateInValidExecutionStatusByIndex(invalidateFromIndex: number, latestValidHashIndex: number): void {
+    // We need to do a two pass invalidation:
+    //   1. we go up and mark all nodes invalid and then
+    //   2. we need do iterate down and mark all children of invalid nodes invalid
+    //
+    // Pass 1, mark all parents invalid
+    /** invalidateAll is a flag which indicates detection of consensus failure */
+    let invalidateAll = false;
+    let invalidateIndex: number | undefined = invalidateFromIndex;
+    while (invalidateIndex !== undefined && invalidateFromIndex > latestValidHashIndex) {
```

Comment:
should it be 
```typescript
while (invalidateIndex !== undefined && invalidateIndex > latestValidHashIndex) {
```

Reply:
I suppose adding some unit tests for those new methods should help

Reply:
thanks for spotting it @tuyennhv ! yes will add a test case around this as well.